hbase-client/src/main/java/org/apache/hadoop/hbase/client/RawAsyncTableImpl.java

```diff
@@ -796,52 +797,102 @@ private boolean locateFinished(RegionInfo region, byte[] endKey, boolean endKeyI
 
   private <S, R> void coprocessorServiceUntilComplete(Function<RpcChannel, S> stubMaker,
     ServiceCaller<S, R> callable, PartialResultCoprocessorCallback<S, R> callback,
-    AtomicBoolean locateFinished, AtomicInteger unfinishedRequest, RegionInfo region, Span span) {
+    MultiRegionCoprocessorServiceProgress<R> progress, RegionInfo region, Span span) {
     addListener(coprocessorService(stubMaker, callable, region, region.getStartKey()), (r, e) -> {
       try (Scope ignored = span.makeCurrent()) {
-        if (e != null) {
-          callback.onRegionError(region, e);
-        } else {
-          callback.onRegionComplete(region, r);
-        }
+        if (e instanceof RegionNameChangedException) {
+          RegionInfo newRegion = ((RegionNameChangedException)e).getNewRegionInfo();
+          if (progress.markNewRegionAndCheckNeedsToBeHandled(newRegion)) {
+            if (progress.hasResponseStarted(region)) {
+              // already started sending responses for the original region, so we can't switch to
+              // the new regions because that would violate the contract of the callbacks
+              progress.onResponse(region, null, new DoNotRetryIOException(
+                "Region " + region.getEncodedName() + " no longer exists, likely due to a split or"
+                  + " merge, and the coprocessor service was already in progress and can't be recovered"
+              ));
+            } else {
+              LOG.debug("Attempted to send a coprocessor service RPC to region {} which no"
+                  + " longer exists, will attempt to send RPCs to the region(s) that replaced it",
+                region.getEncodedName());
+              restartCoprocessorServiceForRange(stubMaker, callable, callback, region.getStartKey(),
```

Comment:
This is where this is submitting a new call for the range covered by the region that was split/merged